File: simpliMvvm/src/main/java/com/trayis/simplimvvm/utils/SimpliViewModelProvidersFactory.java
package com.trayis.simplimvvm.utils;

import android.content.Context;

import java.util.HashMap;
import java.util.Map;

import androidx.lifecycle.ViewModelProvider;

public abstract class SimpliViewModelProvidersFactory implements ViewModelProvider.Factory {

    private Map<Class<? extends SimpliResource>, SimpliResource> resourcesMap = new HashMap<>();

    protected Context context;

    protected void putResource(SimpliResource resource) {
        resourcesMap.put(resource.getClass(), resource);
    }

    @SuppressWarnings("unchecked")
    protected <T extends SimpliResource> T getResource(Class<T> resourceClass) {
        return (T) resourcesMap.get(resourceClass);
    }

}

File: simpliKMvvm/src/main/java/com/trayis/simplikmvvm/utils/SimpliViewModelProvidersFactory.kt
package com.trayis.simplikmvvm.utils

import android.content.Context
import androidx.lifecycle.ViewModelProvider

/**
 * @author mudesai (Mukund Desai)
 * @created on 12/21/18
 */
abstract class SimpliViewModelProvidersFactory : ViewModelProvider.Factory {

    var context: Context? = null

    val resourcesMap = HashMap<Class<SimpliResource>, SimpliResource>()

    fun putResource(resource: SimpliResource) {
        resourcesMap.put(resource.javaClass, resource)
    }

    fun <T : SimpliResource> getResource(resourceClass: Class<in T>) = resourcesMap.get(resourceClass)

}